### ISSUE TITLE
 [A11y] iOS VoiceOver can't move cursor to the end of text field if semantics of the text field is not the same as its content.

### ISSUE THREAD

**Issue by chunhtai** [flutter/flutter#67754]
b/170461718

## Steps to Reproduce

```dart
import 'package:flutter/material.dart';
void main() => runApp(MyApp());

const _spaceCodeUnit = 32;

extension StringSemanticsExtension on String {
  String get letterByLetterSemanticsLabel {
    if (this == null) return null;

    final buffer = StringBuffer();
    final length = this.length;
    for (var i = 0; i < length; ++i) {
      buffer..writeCharCode(_spaceCodeUnit)..writeCharCode(codeUnitAt(i));
    }
    final result = buffer.toString();
    print(result);

    return result;
  }
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final _controller = TextEditingController(text:'424');
  String _currentText = '424';
  bool _announceLetterByLetter = true;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title, style: TextStyle(fontFamily: 'ProductSans')),
      ),
      body: Column(
        children: [
          MergeSemantics(
            child: Semantics(
              label: 'Enter phone number',
              value: _textSemantics,
              // textField: true,
              child: TextField(
                controller: _controller,
                decoration: InputDecoration(hintText: 'Enter text here'),
                keyboardType: TextInputType.phone,
                onChanged: (text) => setState(() => _currentText = text),
              ),
            ),
          ),
          Text('Semantics: $_textSemantics'),
          SwitchListTile(
            value: _announceLetterByLetter,
            onChanged: (state) =>
                setState(() => _announceLetterByLetter = state),
            title: Text('Should announce letter by letter'),
          )
        ],
      ),
    );
  }

  String get _textSemantics {
    if (_announceLetterByLetter) {
      return _currentText.letterByLetterSemanticsLabel;
    }
    return _currentText;
  }
}

```
1. turn on ios voice over
2. focus the textfield and double tap to activate
3. cursor is move to the begining
4. if turn off the ration button, the cursor will be move to the end correctly.

expect: the cursor should always move to the end regardlessly

**Comment by chunhtai** [comment]
it looks like this can only be solved entirely in iOS 13+
there is a new api to make voice over announce the string word by word

In Accessibility_text_entry.mm TextInputSemanticsObject
```mm
- (NSAttributedString*)accessibilityAttributedValue {
  if (![self isAccessibilityBridgeAlive])
    return nil;
  NSString* value = [super accessibilityValue];
  if (value == nil)
    value = [self textInputSurrogate].accessibilityValue;
  if (@available(iOS 13, *)) {
    // Use iOS 13 APIs.
    return [[NSMutableAttributedString alloc] initWithString:value
                                 attributes:@{ UIAccessibilitySpeechAttributeSpellOut : @(YES) }];
  } else {
    return [[NSMutableAttributedString alloc] initWithString:value];
  }
}
```

Waiting for custom reply to see if it is ok to only fix this for ios 13+

**Comment by chunhtai** [comment]
Lower the priority to P3 until we hear back from customer.

**Comment by chunhtai** [comment]
The fix will be as following:

We will add a new parameter in Semantics widget called valueAttribute. which will be an enum {none, spellOut}

In iOS if the value attribute is spellOut, it will override the accessibilityAttributedValue.
In android,  if the value attribute is spellOut it will insert space in between character.

**Comment by chunhtai** [comment]
unassign myself until I have time to get back to this

**Comment by chunhtai** [comment]
@mehmetf isn't the internal p2 translate to our p3?

**Comment by mehmetf** [comment]
No, they are 1:1. Updated the bug.

**Comment by chunhtai** [comment]
API is added closing this issue
